node/test/circuits.js

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var debugLogtron = require('debug-logtron');
+var RelayNetwork = require('./lib/relay_network.js');
+var Circuits = require('../circuits');
+var States = require('../states');
+var MockTimers = require('time-mock');
+
+RelayNetwork.test('should switch to unhealthy', {
```

Comment:
We need more tests.

 - Ensure that circuit breaking does not interfere with cancel & bad request
 - Ensure that circuit breaking works for other exceptions like Timeout and Network
 - Ensure that circuits do not interfere with each other. i.e. multiple service names
 - Ensure that circuits do not interfere with each other. i.e. multiple cn
 - Ensure that circuits do not interfere with each other. i.e. multiple endpoints

You are adding a ton of proxies, closures and hash table lookups in the critical path; I will rip through all this code in my perf work.

I strongly recommend you write a bunch of tests to gaurd us from regressions.